### PR TITLE
cmake: FindPythonInterp: Prioritize 'python' from path the highest

### LINKED ISSUE #11857: cmake does not detect most recent python, it is fixed somehow to 3.4.x
**Describe the bug**
On a system with multiple versions of python where the binary **python** actually points to python 3.7, cmake still thinks I want 3.4.x.

This is hardcoded in the boilerplate, when I change:

```diff
diff --git a/cmake/app/boilerplate.cmake b/cmake/app/boilerplate.cmake
index d17e3e20e0..907881072f 100644
--- a/cmake/app/boilerplate.cmake
+++ b/cmake/app/boilerplate.cmake
@@ -86,10 +86,7 @@ set_property(DIRECTORY APPEND PROPERTY CMAKE_CONFIGURE_DEPENDS ${AUTOCONF_H})
 # that we need to work around until we upgrade to 3.13. Until then we
 # maintain a patched copy in our repo. Bug:
 # https://github.com/zephyrproject-rtos/zephyr/issues/11103
-set(PythonInterp_FIND_VERSION 3.4)
-set(PythonInterp_FIND_VERSION_COUNT 2)
 set(PythonInterp_FIND_VERSION_MAJOR 3)
-set(PythonInterp_FIND_VERSION_MINOR 4)
 set(PythonInterp_FIND_VERSION_EXACT 0)
 set(PythonInterp_FIND_REQUIRED 1)
 include(${ZEPHYR_BASE}/cmake/backports/FindPythonInterp.cmake)
```

it detect the version I want.

**To Reproduce**
Install a new version of python, for example in a virtualenv and make it default

**Expected behavior**
cmake should detect the enabled python configured, not a specific fixed version.

**Impact**
You might have installed all requirements using one version that cmake will not detect and build will fail.

**Additional context**
Some CI systems come with lots of python versions installed, this is confusing and doomed to fail if we pick a specific version.

### PR DESCRIPTION
**Pull request by SebastianBoe**
Users expect that not just "some" compatible python is detected, but
also that it will be the 'python' executable from PATH, at least when
this executable is valid.

To this end rewrite FindPythonInterp to give this executable the
highest priority.

This fixes #11857

Signed-off-by: Sebastian Bøe <sebastian.boe@nordicsemi.no>

### CODE REVIEW

**Review by ulfalizer — CHANGES_REQUESTED** [review]
Can we just error out if `sys.version_info` doesn't exist instead? It was added in Python 2.0, released about 18 years ago.

Would make this code a lot easier to read for CMake noobs like me, without any drawbacks in practice.

File: cmake/backports/FindPythonInterp.cmake [hunk 1]
```diff
@@ -36,6 +36,45 @@
 # get the currently active Python version by default with a consistent version
 # of PYTHON_LIBRARIES.
 
+function(determine_python_version exe result)
+    execute_process(COMMAND "${exe}" -c
+                            "import sys; sys.stdout.write(';'.join([str(x) for x in sys.version_info[:3]]))"
```
[the remark this answers is not in the document]
Comment (SebastianBoe) [reply cmake/backports/FindPythonInterp.cmake]: > Can use a generator expression: ';'.join(str(x) for x in foo)' works the same as ';'.join([str(x) for x in foo]).

This is out-of-scope of the PR. This code has only been moved, not modified, so it's contents are not up for review.

**Comment by SebastianBoe** [comment]
> Can we just error out if sys.version_info doesn't exist instead? It was added in Python 2.0, released about 18 years ago.
> 
> Would make this code a lot easier to read for CMake noobs like me, without any drawbacks in practice.

Fixed.

**Review by ulfalizer — APPROVED** [review]
Code for ancient Python versions removed

**Comment by SebastianBoe** [comment]
> Can we just error out if sys.version_info doesn't exist instead? It was added in Python 2.0, released about 18 years ago.
-- @ulfalizer 

@ulfalizer : Evidently, users have 19 year old python versions installed: https://github.com/zephyrproject-rtos/zephyr/issues/12066#issuecomment-447271357

**Pull request merged**